### PR TITLE
Guards parsing

### PR DESCRIPTION
**Pull request by reilem**

### CODE REVIEW

**Comment by reilem** [comment]
There are still a few issues:

- Parsing “some” with comma separated items?
  - Now using `sepBy1`
- If there is a HsGuarded (a guarded RHS) without bar, and no guards, then we need to disallow further parsing.
- If there is a HsGuarded with bar and guards, then all further HsGuarded must also be of this format.

**Comment by reilem** [comment]
Above issues are hopefully now fixed

File: Frontend/HsParser.hs [hunk 1]
```diff
@@ -281,12 +281,33 @@ pTransPat (PsdAppPat p1 p2) = do
   p2' <- pTransPat p2
   return $ HsConPat dc (xs ++ [p2']) -- NOTE: this is O(n^2) because of the (++), could be O(n)
 
+-- | Wrap the contents of a monad into a singleton array
+one :: Monad f => f a -> f [a]
+one x = x >>= (\x' -> return [x'])
```
Comment (gkaracha) [inline Frontend/HsParser.hs]: Alternatively:
```suggestion
one x = pure <$> x
```
which means that you can inline this function completely :wink:

File: Frontend/HsTypes.hs [hunk 2]
```diff
@@ -119,7 +119,17 @@ data Term a = TmVar (HsTmVar a)                   -- ^ Term variable
 type PsTerm = Term Sym
 type RnTerm = Term Name
 
-data HsAlt a = HsAlt (HsPat a) (Term a)
+data HsGuard   a = HsGuard (HsPat a) (Term a)
```
Comment (gkaracha) [inline Frontend/HsTypes.hs]: Maybe rename the data constructor to `HsPatGuard`? We would like to (eventually) have boolean guards as well

File: Frontend/HsParser.hs [hunk 3]
```diff
@@ -281,12 +281,33 @@ pTransPat (PsdAppPat p1 p2) = do
...
+pGuard =  HsGuard <$> pPat <* symbol "<-" <*> pTerm
+
+-- | Parse an arrow and right hand side term
+pRhs :: PsM PsTerm
+pRhs = symbol "->" *> pTerm
+
+-- | Parse a list of guarded right hand sides
+pGuardedRhsList :: PsM [PsGuarded]
+pGuardedRhsList =  one  (HsGuarded <$> pure [] <*> pRhs)
+               <|> some (HsGuarded <$> pGuards <*> pRhs)
```
Comment (gkaracha) [inline Frontend/HsParser.hs]: I find this definition weird

File: Frontend/HsParser.hs [hunk 4]
```diff
@@ -281,12 +281,33 @@ pTransPat (PsdAppPat p1 p2) = do
...
+pGuards :: PsM [PsGuard]
+pGuards = indent (symbol "|" *> sepBy1 pGuard (symbol ", "))
+
+-- | Parse a list of pattern guards
+pGuard :: PsM PsGuard
+pGuard =  HsGuard <$> pPat <* symbol "<-" <*> pTerm
+
+-- | Parse an arrow and right hand side term
+pRhs :: PsM PsTerm
+pRhs = symbol "->" *> pTerm
```
Comment (gkaracha) [inline Frontend/HsParser.hs]: Best if you inline this definition. I know it looks repetitive but it doesn't make much sense as a standalone function.

**Review by gkaracha — COMMENTED** [review]
Just dropped some comments. Let's discuss this PR more tomorrow.

File: Frontend/HsParser.hs (hunk 1 above)
Comment (reilem) [reply Frontend/HsParser.hs]: Damnit I searched so hard to make it smallest but I knew there was going to be something shorter that I had missed haha 😅

File: Frontend/HsTypes.hs [hunk 5]
```diff
@@ -142,6 +152,18 @@ instance (Symable a, PrettyPrint a) => PrettyPrint (HsAlt a) where
...
   needsParens _      = True
 
+instance (Symable a, PrettyPrint a) => PrettyPrint (HsPatGuard a) where
+  ppr (HsPatGuard p t) = ppr p <+> text "<-" <+> ppr t
+  needsParens _     = False
+
+instance (Symable a, PrettyPrint a) => PrettyPrint (HsGuarded a) where
+  ppr (HsGuarded [] t) = ppr t
+  ppr (HsGuarded gs t) = text "|"
+    <+> text (intercalate ", " (map (render . ppr) gs))
```
Comment (gkaracha) [inline Frontend/HsTypes.hs]: Use only combinators if you can; don't go to String and then back to Doc. That is, do this instead:
```suggestion
    <+> fsep (punctuate comma (map ppr gs))
```

File: Frontend/HsParser.hs [hunk 6]
```diff
@@ -281,12 +281,30 @@ pTransPat (PsdAppPat p1 p2) = do
...
+
 -- | Parse a pattern
 pPat :: PsM PsPat
 pPat = pTransPat <$> pPatMain >>= \case
   Nothing -> empty
   Just p  -> return p
 
+-- | Parse guards
+pGuards :: PsM [PsGuard]
+pGuards = indent (symbol "|" *> sepBy1 pGuard (symbol ", "))
```
Comment (gkaracha) [inline Frontend/HsParser.hs:297]: Replace `", "` with `","`; I assume that the space there is by accident.

**Pull request merged**